lib/std/smartptrs.nim

```diff
@@ -0,0 +1,130 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2021 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+
+## C++11 like smart pointers. They always use the shared allocator.
+import std/isolation
+
+type
+  UniquePtr*[T] = object
+    ## Non copyable pointer to a value of type `T` with exclusive ownership.
+    val: ptr T
+
+proc `=destroy`*[T](p: var UniquePtr[T]) =
+  if p.val != nil:
+    `=destroy`(p.val[])
+    deallocShared(p.val)
+
+proc `=copy`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.error.}
+  ## The copy operation is disallowed for `UniquePtr`, it
+  ## can only be moved.
+
+proc newUniquePtr[T](val: sink Isolated[T]): UniquePtr[T] {.nodestroy.} =
+  ## Returns a unique pointer which has exclusive ownership of the value.
+  result.val = cast[ptr T](allocShared(sizeof(T)))
+  # thanks to '.nodestroy' we don't have to use allocShared0 here.
+  # This is compiled into a copyMem operation, no need for a sink
+  # here either.
+  result.val[] = extract val
+  # no destructor call for 'val: sink T' here either.
+
+template newUniquePtr*[T](val: T): UniquePtr[T] =
+  ## .. warning:: Using this template in a loop causes multiple evaluations of `value`.
+  newUniquePtr(isolate(val))
+
+proc isNil*[T](p: UniquePtr[T]): bool {.inline.} =
+  p.val == nil
+
```

Comment:
https://github.com/nim-lang/Nim/pull/17333#issuecomment-803887085 is still very much relevant (please read the whole comment)

Reply:
no need for trickery see this comment: https://github.com/nim-lang/Nim/pull/17333#issuecomment-870486146

Reply:
the example in
> UniquePtr should work with importcpp types but currently this gives weird results:

doesn't work. Even after trying several modifications and adapting to new API (eg get=>[] etc); several errors:
* `checkNotNil` gives a Error: type mismatch error
* then i get a `error: member reference type 'Foo *' is a pointer; did you mean to use '->'?` error and other errors.

smartptrs need to work with importcpp types, it's an important use case.

Reply:
I can add an empty constructor overload, if that's all it takes

Reply:
what output will it give? right now it gives:
```nim
ok1
ctor2: 3
ctor: 0
dtor: 3
ctor: 0
dtor: 3
dtor: 0
dtor: 3
ok2
3
ok3
dtor: 3
ok4
```
which doesn't look correct

Reply:
> so it doesn't seem like relying on backend compiler is reliable then?

The example is pretty extreme, and in normal use a copyMem doesnt really matter. however to be absolutely certain this wont happen I am offering to add a more manual (and slower) construction method of first creating a zero initialized location, then constructing your object with []= I expect that to work with all compilers.

So If I understand correctly you are getting compiler errors, Is there something I can do, in the scope of smartptrs to fix them?

Reply:
adding this will solve a lot of problems:
```nim
template newUniquePtr*[T](U: typedesc[T]): UniquePtr[T] =
  var result: UniquePtr[T]
  result.val = cast[ptr T](allocShared(sizeof(T)))
  result
```
the C++ example now gives:
```nim
ok1
ok2
0
ok3
dtor: 0
ok4
```
and user code can probably be instrumented to add placement-new, which will get closer to C++ semantics of unique_ptr and avoid duplicate constructors (as well as for the above mentioned issue regarding double allocation on stack then copy on heap)

Reply:
> (as well as for the above mentioned issue regarding double allocation on stack then copy on heap)

What I am trying to point out though... is that there is at least a modern compiler that can optimize this to what people expect. Imo we cannot bother offering anything similar to placement new, like emplace_back for seqs. Hopefully the other compilers will catch up.

Reply:
> proc newUniquePtr*[T](U: typedesc[T]): UniquePtr[T] =
  result.val = cast[ptr T](allocShared(sizeof(T)))

Either I call it `newUniquePtr` + `Uninitialized` or I should use `allocShared0`. What's preferred?